core/src/main/java/com/linecorp/armeria/client/Proxy.java

```diff
@@ -0,0 +1,209 @@
+ */
+public class Proxy {
+
+    private static final long USE_DEFAULT_TIMEOUT_MILLIS = -1;
+    static final Proxy DEFAULT =
+            new Proxy(ProxyType.NONE, new InetSocketAddress(0), USE_DEFAULT_TIMEOUT_MILLIS);
+
+    private final ProxyType proxyType;
+    private final InetSocketAddress proxyAddress;
+    private final long connectTimeoutMillis;
+
+    @Nullable
+    private String userName;
+
+    Proxy(ProxyType proxyType, InetSocketAddress proxyAddress, long connectTimeoutMillis) {
+        this.proxyType = proxyType;
+        this.proxyAddress = proxyAddress;
+        this.connectTimeoutMillis = connectTimeoutMillis;
+    }
+
+    /**
+     * Creates a {@code Proxy} configuration for SOCKS4 protocol.
+     * @param proxyAddress The proxy address.
+     */
+    public static Socks4ProxyBuilder socks4(InetSocketAddress proxyAddress) {
+        return new Socks4ProxyBuilder(requireNonNull(proxyAddress), USE_DEFAULT_TIMEOUT_MILLIS);
+    }
+
+    /**
+     * Creates a {@code Proxy} configuration for SOCKS4 protocol.
+     * @param proxyAddress The proxy address.
+     * @param connectTimeoutMillis The connection timeout for connecting to the proxy server.
+     */
+    public static Socks4ProxyBuilder socks4(InetSocketAddress proxyAddress, long connectTimeoutMillis) {
+        checkArgument(connectTimeoutMillis >= 0);
+        return new Socks4ProxyBuilder(requireNonNull(proxyAddress), connectTimeoutMillis);
+    }
+
+    /**
+     * Creates a {@code Proxy} configuration for SOCKS5 protocol.
+     * @param proxyAddress The proxy address.
+     */
+    public static Socks5ProxyBuilder socks5(InetSocketAddress proxyAddress) {
+        return new Socks5ProxyBuilder(requireNonNull(proxyAddress), USE_DEFAULT_TIMEOUT_MILLIS);
+    }
+
+    /**
+     * Creates a {@code Proxy} configuration for SOCKS5 protocol.
+     * @param proxyAddress The proxy address.
+     * @param connectTimeoutMillis The connection timeout for connecting to the proxy server.
+     */
+    public static Socks5ProxyBuilder socks5(InetSocketAddress proxyAddress, long connectTimeoutMillis) {
+        checkArgument(connectTimeoutMillis >= 0);
+        return new Socks5ProxyBuilder(requireNonNull(proxyAddress), connectTimeoutMillis);
+    }
+
+    /**
+     * Creates a {@code Proxy} configuration for CONNECT protocol.
+     * @param proxyAddress The proxy address.
+     */
+    public static ConnectProxyBuilder connect(InetSocketAddress proxyAddress) {
+        return new ConnectProxyBuilder(requireNonNull(proxyAddress), USE_DEFAULT_TIMEOUT_MILLIS);
+    }
+
+    /**
+     * Creates a {@code Proxy} configuration for CONNECT protocol.
+     * @param proxyAddress The proxy address.
+     * @param connectTimeoutMillis The connection timeout for connecting to the proxy server.
+     */
+    public static ConnectProxyBuilder connect(InetSocketAddress proxyAddress, long connectTimeoutMillis) {
```

Comment:
if connect timeout is optional, we can remove the factory methods with `connectTimeoutMillis` and add `connectTimeout(Duration)` and `connectTimeoutMillis(long)` to the builder class.

Reply:
Decided to get rid of proxy specific `connectTimeoutMillis ` and use `ClientFactory.connectTimeoutMillis`.